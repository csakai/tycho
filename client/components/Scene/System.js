import Scene from './Scene';
import Fixtures from 'global/fixtures';
import Orbital from './Objects/Orbital';
import Clock from 'engine/clock';

export default class System {

  constructor(data) {
    this.data = Fixtures;//data;//temp
    this.clock = new Clock();
    this.setUp();
    this.animate();
  }

  /**
   * Setup.
   */
  setUp = () => {
    this.renderSkybox();
    this.renderScene();
    this.renderOrbitals();
    this.renderSun();
  }

  /**
   * Renders the skybox (Milky Way)
   */
  renderSkybox = () => {
    // TODO
  }

  /**
   * Renders the Sun.
   */
  renderSun = () => {
    // TODO
  }

  /**
   * Renders the scene.
   */
  renderScene = () => {
    this.scene = new Scene(this.clock);
  }

  /**
   * Renders the orbitals from the data.
   */
  renderOrbitals = () => {
    this.orbitals = this.data.map(this.addOrbital);
  }

  /**
   * Animation loop.
   */
  animate = () => {
    this.scene.animate();
    this.orbitals.forEach(this.updateOrbital);
    this.clock.update();
    requestAnimationFrame(this.animate);
  }

  /**
   * Updates the positions of an orbital.
   * @param  {Orbital} orbital
   */
  updateOrbital = (orbital) => {
    orbital.updatePosition(this.clock.elapsedTime);
  }

  /**
   * Adds an orbital to the scene. The orbital's angle of
   * inclination is raised to the ecliptic plane.
   * @param  {Object} orbital - data of orbital to add
   * @return {Orbital} instance of orbital added
   */
  addOrbital = (orbital) => {
    orbital.inclination -= 90;
    return new Orbital(orbital, this.scene);
  }

  /**
   * Events to execute on each clock second ("tick")
   * @param  {Function} fn function to execute
   */
  tick = (fn) => {
    this.clock.tick(fn);
  }

  /**
   * Zooms to specified level
   * @param  {Number} level zoom percentage [0,1]
   */
  zoom = (level) => {
    this.scene.controls.zoom(level);
  }

  /**
   * Change the proportion (scale) of the props in scene
   * @param  {Number} proportion scalar to multiply sizes by,
   *                             as an exponent of 10
   */
  size = (scale) => {
    // Math.pow(10, scale);
  }

  /**
   * Change the scale of the props in scene
   * @param  {Number} speed scalar to multiply time elapsed by,
   *                        as an exponent of 10
   */
  speed = (speed) => {
    this.clock.speed(speed);
  }
}
