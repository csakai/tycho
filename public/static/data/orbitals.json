[{
	"name": "Earth",
	"albedo": 0.367,
	"sidereal": 0.99726968,
	"radius": 6371,
	"markdown": "earth.md",
	"mass": 5.97237e+24,
	"centralMass": 1.989e+30,
	"axialTilt": 23.4392811,
	"inclination": 7.155,
	"atmosphere": 35260,
	"satellites": [{
		"name": "Moon",
		"albedo": 0.136,
		"sidereal": 0.99726968,
		"radius": 1737.1,
		"markdown": "moon.md",
		"mass": 5.97237e+24,
		"centralMass": 5.97237e+24,
		"axialTilt": 6.687,
		"inclination": 5.145,
		"atmosphere": 10066278,
		"ephemeris": {
			"nasaOrbitalCode": "301",
			"nasaBarycenterCode": "3",
			"revolutionOrder": "d"
		},
		"argumentOfPeriapsis": 282.5992409518417,
		"eccentricity": 0.03405532141489297,
		"longitudeOfAscendingNode": 142.3595722910625,
		"semimajor": 376126.5103019063,
		"semiminor": 375908.33786092995,
		"periapses": {
			"last": 1509927058416.626,
			"next": 1512377208118.1555
		},
		"id": "moon"
	}],
	"ephemeris": {
		"nasaOrbitalCode": "3",
		"nasaBarycenterCode": "0",
		"revolutionOrder": "y"
	},
	"argumentOfPeriapsis": 114.20763,
	"eccentricity": 0.01355360061112275,
	"longitudeOfAscendingNode": 348.73936,
	"semimajor": 150759239.9570005,
	"semiminor": 150745392.07806867,
	"periapses": {
		"last": 1136430000000,
		"next": 1167987600000
	},
	"id": "earth"
},{
	"name": "Venus",
	"albedo": 0.689,
	"sidereal": -243.025,
	"radius": 6051.8,
	"markdown": "venus.md",
	"mass": 4.8675e+24,
	"centralMass": 1.989e+30,
	"axialTilt": 2.64,
	"inclination": 3.86,
	"atmosphere": 16052641,
	"ephemeris": {
		"nasaOrbitalCode": "2",
		"nasaBarycenterCode": "0",
		"revolutionOrder": "y"
	},
	"argumentOfPeriapsis": 55.186,
	"eccentricity": 0.0067,
	"longitudeOfAscendingNode": 76.678,
	"semimajor": 108208179,
	"semiminor": 107477094,
	"periapses": {
		"last": 1429348560000,
		"next": 1448787300000
	},
	"id": "venus"
}]