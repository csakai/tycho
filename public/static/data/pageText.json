{"planetScale":"Planet Scale: × ","speedScale":"Speed Scale: × 10^","settings":"Settings"}
