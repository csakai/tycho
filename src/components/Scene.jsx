import React from 'react';
import PropTypes from 'prop-types';
import OrbitalContainer from '../containers/OrbitalContainer';
import Skybox from './Skybox';

export default class Scene extends React.Component {

  static propTypes = {
    orbitalData: PropTypes.array.isRequired,
    updatePosition: PropTypes.func.isRequired,
    time: PropTypes.number,
    camera: PropTypes.object,
    scale: PropTypes.number
  }

  getOrbitalElements = (orbitals, isSatellite) => {
    return orbitals.map((orbital) => (
      <OrbitalContainer
        {...orbital}
        scale={this.props.scale}
        time={this.props.time}
        camera={this.props.camera}
        updatePosition={this.props.updatePosition}
        isSatellite={isSatellite}
        key={orbital.id}>
        {orbital.satellites && this.getOrbitalElements(orbital.satellites, !isSatellite)}
      </OrbitalContainer>
    ));
  }

  render() {
    return (
      <group>
        {this.getOrbitalElements(this.props.orbitalData)}
        <axisHelper size={500} />
        <pointLight
          color={0xffffff}
          intensity={0.95}
          distance={500}
        />
        <Skybox />
        <mesh lookAt={this.props.cameraMatrix}>
          <planeGeometry width={100} height={100} />
          <meshBasicMaterial color={0x00ff00} />
        </mesh>
      </group>
    );
  }
}
