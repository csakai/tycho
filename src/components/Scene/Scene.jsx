import React from 'react';
import PropTypes from 'prop-types';
import OrbitalContainer from '../../containers/OrbitalContainer';
import Sun from '../Sun';

export default class Scene extends React.Component {

  static propTypes = {
    orbitalData: PropTypes.array.isRequired,
    time: PropTypes.number,
    camera: PropTypes.object,
    scale: PropTypes.number,
    domEvents: PropTypes.object,
    highlightedOrbitals: PropTypes.array,
    targetId: PropTypes.string
  }

  getOrbitalElements = (orbitals, isSatellite) => {
    return orbitals.map((orbital) => (
      <OrbitalContainer
        {...orbital}
        key={orbital.id}
        action={this.props.action}
        scale={this.props.scale}
        time={this.props.time}
        camera={this.props.camera}
        domEvents={this.props.domEvents}
        targetId={this.props.targetId}
        highlightedOrbitals={this.props.highlightedOrbitals}
        isSatellite={isSatellite}>
        {orbital.satellites && this.getOrbitalElements(
          orbital.satellites, !isSatellite
        )}
      </OrbitalContainer>
    ));
  }

  render() {
    return (
      <group>
        {this.getOrbitalElements(this.props.orbitalData)}
        <Sun  camera={this.props.camera} />
      </group>
    );
  }
}
