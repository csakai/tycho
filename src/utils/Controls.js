import * as THREE from 'three';
import OrbitControls from 'three-orbit-controls';
import TWEEN from 'tween.js';
import Constants from '../constants';

export default class Controls extends OrbitControls(THREE) {

  constructor(camera, domElement) {
    super(camera, domElement);

    this.camera = camera;
    this.enabled = true;
    this.enableZoom = false;
    this.enablePan = false;
    this.level = Constants.WebGL.Zoom.MAX;
    this.minDistance = Constants.WebGL.Camera.MIN_DISTANCE;
    this.maxDistance = Constants.WebGL.Camera.MAX_DISTANCE;
  }

  /**
   * Change zoom. Overrides the existing zoom tween in progress.
   *
   * @param {level} level - percentage of desired zoom level [0,100]
   */
  zoom = (level) => {
    if(this.level !== level) {
      this.pan(level / 100);
      this.level = level;
    }
  }

  /**
   * Gets the zoom delta from controls and invokes the given action if zoom has changed.
   *
   * @param {Event} ev - DOM wheel event
   * @param {Function} action - callback action
   */
  wheelZoom = (ev, action) => {
    const zoom = this.getZoomDelta(ev.deltaY);
    const current = Math.round(this.level * 100);

    if (current !== zoom) {
      action(zoom);
    }
  }

  /**
   * Calculates the zoom percentage from the given mousewheel delta.
   *
   * @param {Number} delta - mousewheel delta
   * @returns {Number} new zoom level
   */
  getZoomDelta = (delta) => {
    let zoom = this.level;
    
    zoom += (delta / this.getWheelDeltaDivisor(zoom));
    zoom = Math.max(zoom, 0);
    zoom = Math.min(zoom, 100);
   
    return zoom;
  }

  /**
   * Calculates a wheel delta that inflates based on how far away
   * the user is from the target, because outer space is really big.
   *
   * @param {Number} level - current zoom level 
   * @returns {Number} scrollwheel delta
   */
  getWheelDeltaDivisor = (level) => {
    let delta = Constants.UI.WHEEL_DELTA_DIVISOR;
    let match = false;

    Constants.WebGL.ScrollScale
      .forEach(({distance, scale}) => {
        if (level > distance && !match) {
          delta *= scale;
          match = true;
        }
      });

    return delta;
  }

  /**
   * Sets the current camera position to the scaled zoom vector.
   *
   * @param {Number} percent - percentage of zoom [0,1]
   */
  pan = (percent, log) => {
    let position = this.camera.position;
    let newVector = this.getZoomVector(position, this.maxDistance * percent);
    let minVector = this.getZoomVector(position, this.minDistance);
    
    if (newVector.length() >= this.minDistance) {
      position.copy(newVector);
    } else {
      position.copy(minVector);
    }
  }

  /**
   * Returns the current vector scaled to the desired magnitude.
   *
   * @param {Vector3} vector - current position vector
   * @param {Number} scalar - magnitude of vector
   * @returns {Vector3}
   */
  getZoomVector = (vector, scalar) => {
    return vector
      .clone()
      .normalize()
      .multiplyScalar(scalar);
  }

  /**
   * Returns the distance between <0> and current camera position.
   * 
   * @returns {Number} distance
   */
  getDistance = () => {
    return this.maxDistance - this.minDistance;
  }

  /**
   * Enable controls.
   */
  enable = () => {
    this.enabled = true;
  }

  /**
   * Disable controls.
   */
  disable = () => {
    this.enabled = false;
  }

  /**
   * Starts autorotating the scene clockwise.
   *
   * @param {Number} speed - speed to rotate at
   */
  startAutoRotate = (speed) => {
    this.autoRotate = true;
    this.autoRotateSpeed = speed;
  }

  /**
   * Stop autorotating the scene.
   */
  stopAutoRotate = () => {
    this.autoRotate = false;
  }

  /**
   * Deletes Tween instance and data.
   */
  endTween = () => {
    delete this.tweenData;
    delete this.tweenBase;
  }

  /**
   * Cancels Tween, if one is in progress.
   */
  cancelTween = () => {
    if (this.tweenBase) {
      this.tweenBase.stop();
      this.endTween();
    }
  }

  /**
   * Zooms to the level of the current Tween data.
   */
  updateTween = () => {
    this.zoom(this.tweenData.level);
  }

  /**
   * Invokes teardown methods for active tween.
   */
  completeTween = () => {
    if (this.tweenDone) {
      this.tweenDone(this.level);
    }
    this.endTween();
  }

  /**
   * Creates an animation to zoom to the specified level.
   *
   * @param {Number} level - percentage of zoom
   * @param {Function} onDone - animation completion callback
   */
  tweenZoom = (level, onDone) => {
    this.cancelTween();

    this.tweenDone = onDone;
    this.tweenData = {level: this.level};

    this.tweenBase = new TWEEN
      .Tween(this.tweenData)
      .easing(TWEEN.Easing.Quadratic.Out)
      .to({level}, Constants.WebGL.Tween.SLOW)
      .onUpdate(this.updateTween)
      .onComplete(this.completeTween)
      .start();
  }
}
