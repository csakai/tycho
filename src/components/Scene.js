import * as THREE from 'three';
import React from 'react';
import PropTypes from 'prop-types';
import React3 from 'react-three-renderer';
import Controls from '../utils/Controls';
import OrbitalContainer from '../containers/OrbitalContainer';

const cameraPosition = new THREE.Vector3(300, 300, 300);//move to const

export default class Scene extends React.Component {

  static propTypes = {
    orbitalData: PropTypes.array.isRequired,
    onAnimate: PropTypes.func.isRequired,
    updateScreenPositions: PropTypes.func.isRequired,
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,
    time: PropTypes.number
  }

  componentDidMount = () => {
    this.controls = new Controls(this.refs.camera);
  }

  componentWillUnmount = () => {
    this.controls.dispose();
    delete this.controls;
  }

  getCamera = (width, height) => {
    return (
      <perspectiveCamera
        name="camera"
        ref="camera"
        fov={50}
        aspect={width / height}
        near={1}
        far={10000}
        position={cameraPosition}
      />
    );
  }

  getCameraPosition = (camera) => {
    if(camera) {
      return camera.position.clone();
    }
    return null;
  }

  getOrbitalElements = (orbitals, odd) => {
    return orbitals.map((orbital) => (
      <OrbitalContainer
        {...orbital}
        time={this.props.time}
        camera={this.refs.camera}
        onUpdate={this.props.updateScreenPositions}
        odd={odd}
        key={orbital.id}>
        {orbital.children && this.getOrbitalElements(orbital.children, !odd)}
      </OrbitalContainer>
    ));
  }

  render() {
    const {width, height} = this.props;
    const camera = this.getCamera(width, height);

    return (
      <React3
        onAnimate={this.props.onAnimate}
        mainCamera="camera"
        width={width}
        height={height}
        antialias={true}
        alpha={true}>
        <scene>
          {camera}
          {this.getOrbitalElements(this.props.orbitalData)}
          <axisHelper size={500} />
          <mesh lookAt={this.getCameraPosition(this.refs.camera)}>
            <planeGeometry width={100} height={100} />
            <meshBasicMaterial color={0x00ff00} />
          </mesh>
        </scene>
      </React3>
    );
  }
}
