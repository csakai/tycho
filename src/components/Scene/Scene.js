import * as THREE from 'three';
import React from 'react';
import PropTypes from 'prop-types';
import React3 from 'react-three-renderer';
import Controls from '../../utils/Controls';
import OrbitalContainer from './containers/OrbitalContainer';

export default class Scene extends React.Component {

  static propTypes = {
    orbitalData: PropTypes.array.isRequired,
    onAnimate: PropTypes.func.isRequired,
    updateScreenPositions: PropTypes.func.isRequired,
    time: PropTypes.number
  }

  componentWillMount = () => {
    this.cameraPosition = new THREE.Vector3(300, 300, 300);
  }

  componentDidMount = () => {
    this.controls = new Controls(this.refs.camera);
  }

  componentWillUnmount = () => {
    this.controls.dispose();
    delete this.controls;
  }

  getCamera = (width, height) => {
    return (
      <perspectiveCamera
        name="camera"
        ref="camera"
        fov={50}
        aspect={width / height}
        near={1}
        far={10000}
        position={this.cameraPosition}
      />
    );
  }

  getOrbitalElements = (orbitals, odd) => {
    return orbitals.map((orbital) => (
      <OrbitalContainer
        {...orbital}
        time={this.props.time}
        camera={this.refs.camera}
        onUpdate={this.props.updateScreenPositions}
        odd={odd}
        key={orbital.id}>
        {orbital.children && this.getOrbitalElements(orbital.children, !odd)}
      </OrbitalContainer>
    ));
  }

  render() {
    const width = window.innerWidth; // canvas width
    const height = window.innerHeight; // canvas height
    const camera = this.getCamera(width, height);

    return (
      <React3
        mainCamera="camera" // this points to the perspectiveCamera which has the name set to "camera" below
        width={width}
        height={height}
        antialias={true}
        onAnimate={this.props.onAnimate}
        alpha={true}>
        <scene>
          {camera}
          {this.getOrbitalElements(this.props.orbitalData)}
          <axisHelper size={500} />
        </scene>
      </React3>
    );
  }
}
