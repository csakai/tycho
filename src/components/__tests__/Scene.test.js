import React from 'react';
import React3 from 'react-three-renderer';
import renderer from 'react-test-renderer';
import ReactDOM from 'react-dom';
import {Vector3, Camera} from 'three';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import Scene from '../Scene';
import Controls from '../../utils/Controls';
import data from '../../global/fixtures';

describe('Scene Component', () => {
  let component, scene;

  beforeEach(() => {
    component = shallow(<Scene
      orbitalData={data}
      updatePosition={() => {}}
      time={1}
    />);

    scene = component.instance();
  });

  describe('getCameraPosition()', () => {
    it('should returned the cloned position vector', () => {
      const position = new Vector3(1, 2, 3);
      const camera = {position};
      const cloned = scene.getCameraPosition(camera);

      expect(cloned !== position).toBe(true);
      expect(cloned).toEqual(position);
    });

    it('should return null if the camera is undefined', () => {
      const cloned = scene.getCameraPosition();

      expect(cloned).toBeDefined();
      expect(cloned).toBeNull();
    });
  });

  describe('getOrbitalElements()', () => {
    let orbitalContainers;

    beforeEach(() => {
      orbitalContainers = scene.getOrbitalElements(data);
    });

    it('should be an array', () => {
      expect(Array.isArray(orbitalContainers)).toBe(true)
    });

    it('should match the snapshot of OrbitalElements', () => {
      const orbitalJson = toJson(shallow(
        <group>orbitalContainers</group>
      ));

      expect(orbitalJson).toMatchSnapshot();
    });
  });

  describe('render()', () => {
    it('should render the scene', () => {
      expect(toJson(component)).toMatchSnapshot();
    });
  });
});
