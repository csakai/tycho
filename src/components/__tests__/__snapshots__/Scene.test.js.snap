// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scene Component getOrbitalElements() should match the snapshot of OrbitalElements 1`] = `
<group>
  orbitalContainers
</group>
`;

exports[`Scene Component render() should render the scene 1`] = `
<group>
  <AnimatedComponent
    GM={6836529}
    arcRotate={45.0411}
    argPeriapsis={114.20763}
    atmosphereColor={8421504}
    axialTilt={23.26}
    eccentricity={0.01671123}
    id="dummyOuter"
    inclination={1.57869}
    longAscNode={348.73936}
    name="Dummy Outer"
    periapses={
      Object {
        "last": 1136430000,
        "next": 1167987600,
      }
    }
    radius={637800}
    ring={
      Object {
        "innerRadius": 6630,
        "outerRadius": 950000,
      }
    }
    semimajor={249598261}
    semiminor={249556483}
    time={1}
    updatePosition={[Function]}
  />
  <AnimatedComponent
    GM={6836529}
    arcRotate={45.0411}
    argPeriapsis={114.20763}
    atmosphereColor={8421504}
    axialTilt={23.26}
    eccentricity={0.01671123}
    id="dummyParent"
    inclination={15.57869}
    longAscNode={348.73936}
    name="Dummy Planet"
    periapses={
      Object {
        "last": 1156430000,
        "next": 1167987600,
      }
    }
    radius={637800}
    ring={
      Object {
        "innerRadius": 6630,
        "outerRadius": 950000,
      }
    }
    satellites={
      Array [
        Object {
          "GM": 6836529,
          "arcRotate": 15.0411,
          "argPeriapsis": 114.20763,
          "atmosphereColor": 16711680,
          "axialTilt": 23.26,
          "eccentricity": 0.01671123,
          "id": "dummyChild",
          "inclination": 1.57869,
          "longAscNode": 348.73936,
          "name": "Dummy Satellite",
          "periapses": Object {
            "last": 1166430000,
            "next": 1167987600,
          },
          "radius": 637800,
          "semimajor": 149598261,
          "semiminor": 149556483,
        },
      ]
    }
    semimajor={149598261}
    semiminor={149556483}
    time={1}
    updatePosition={[Function]}
  >
    <AnimatedComponent
      GM={6836529}
      arcRotate={15.0411}
      argPeriapsis={114.20763}
      atmosphereColor={16711680}
      axialTilt={23.26}
      eccentricity={0.01671123}
      id="dummyChild"
      inclination={1.57869}
      longAscNode={348.73936}
      name="Dummy Satellite"
      odd={true}
      periapses={
        Object {
          "last": 1166430000,
          "next": 1167987600,
        }
      }
      radius={637800}
      semimajor={149598261}
      semiminor={149556483}
      time={1}
      updatePosition={[Function]}
    />
  </AnimatedComponent>
  <axisHelper
    size={500}
  />
  <mesh
    lookAt={null}
  >
    <planeGeometry
      height={100}
      width={100}
    />
    <meshBasicMaterial
      color={65280}
    />
  </mesh>
</group>
`;
